File: wayfinder_router/reliability.py
"""Deterministic gateway reliability primitives (WF-ADR-0031, WF-DESIGN-0010).

Retry classification, a same-tier delivery plan, and a per-target circuit breaker —
all computed from observed transport outcomes, with no model call and no effect on the
scored decision (WF-ADR-0001). These change *how a request is delivered*, never *what
was decided*. Pure and clock/rng-injectable so the gateway's forward path stays testable
without a network.
"""

from __future__ import annotations

import random
import sqlite3
import time
from collections.abc import Callable, Iterable
from dataclasses import dataclass, field

# Upstream statuses worth retrying: rate limiting and transient server faults. Ordinary
# 4xx (bad request, auth) is the client's fault and is never retried.
RETRYABLE_STATUS = frozenset({429, 500, 502, 503, 504})

# Statuses that mean *this target is unusable* rather than *this request was bad*: a
# missing/expired/forbidden upstream key. They are not retryable (retrying a bad key is
# pointless), but — unlike an ordinary client 4xx — they must count as a breaker *failure*
# so a stale key eventually opens the breaker and delivery degrades (WF-ADR-0031).
AUTH_FAILURE_STATUS = frozenset({401, 403, 407})


def is_retryable(status: int | None) -> bool:
    """Whether a forward attempt should be retried.

    ``status is None`` means a transport failure (timeout, connection refused) — always
    retryable. An HTTP status is retryable only for rate-limit / transient 5xx.
    """
    return status is None or status in RETRYABLE_STATUS


def is_auth_failure(status: int | None) -> bool:
    """Whether ``status`` means the upstream target is unusable (bad/expired/forbidden key)."""
    return status in AUTH_FAILURE_STATUS


def retry_delays(
    retries: int,
    *,
    base: float = 0.2,
    cap: float = 5.0,
    rng: Callable[[], float] = random.random,
) -> list[float]:
    """Backoff delays (seconds) before each retry: exponential, capped, full-jitter.

    ``retries`` is the number of *re*-tries, so the schedule has that many entries (the
    first attempt has no preceding delay). Jitter is over ``[0, slot]`` via ``rng``,
    which is injectable for deterministic tests.
    """
    delays: list[float] = []
    for i in range(max(0, retries)):
        slot = min(cap, base * (2**i))
        delays.append(round(slot * rng(), 6))
    return delays


@dataclass
class CircuitBreaker:
    """Per-target breaker: open after ``threshold`` consecutive failures; probe after ``cooldown``.

    Pure bookkeeping over success/failure outcomes — no model call. ``clock`` is injectable
    (default ``time.monotonic``) so cooldown is testable. State is in-memory per process by
    default; optionally (WF-DESIGN-0013 §7c, WF-ROADMAP-0012), when ``state_path`` is set each
    target's ``(fails, opened_at)`` is best-effort persisted to a shared ``state.db`` on every
    ``record`` and reloaded on construction, so an open breaker can survive a restart. The
    guarantee is never-raise: any ``sqlite3.Error``/``OSError`` degrades silently to pure
    in-memory operation. LIMITATION: a persisted ``opened_at`` is a raw ``clock()`` reading;
    ``time.monotonic`` restarts from an arbitrary base across a real process restart, so a
    reloaded cooldown is only meaningful relative to the *same* clock (round-trips under an
    injected clock; undefined across a real monotonic restart). Documented, not solved.
    """

    threshold: int = 5
    cooldown: float = 30.0
    clock: Callable[[], float] = time.monotonic
    state_path: str | None = field(default=None, kw_only=True)
    _fails: dict[str, int] = field(default_factory=dict, repr=False)
    _opened_at: dict[str, float] = field(default_factory=dict, repr=False)
    _conn: sqlite3.Connection | None = field(default=None, init=False, repr=False, compare=False)

    def __post_init__(self) -> None:
        """Best-effort open + reload of persisted per-target state; degrade to in-memory."""
        if self.state_path is None:
            return
        try:
            conn = sqlite3.connect(self.state_path, isolation_level=None)
            conn.execute(
                "CREATE TABLE IF NOT EXISTS breaker ("
                "target TEXT PRIMARY KEY, fails INTEGER, opened_at REAL)"
            )
            rows = conn.execute("SELECT target, fails, opened_at FROM breaker").fetchall()
        except (sqlite3.Error, OSError):
            self._conn = None
            return
        self._conn = conn
        for target, fails, opened_at in rows:
            self._fails[target] = int(fails)
            if opened_at is not None:  # NULL opened_at means the target is closed
                self._opened_at[target] = float(opened_at)

    def allow(self, target: str) -> bool:
        """True if ``target`` may be tried now — closed, or cooldown elapsed (half-open probe)."""
        opened = self._opened_at.get(target)
        if opened is None:
            return True
        return self.clock() - opened >= self.cooldown

    def is_open(self, target: str) -> bool:
        """True while ``target`` is tripped and still cooling down (the inverse of allow)."""
        return not self.allow(target)

    def record(self, target: str, ok: bool) -> None:
        """Fold one attempt's outcome in: success closes the breaker; failures may open it."""
        if ok:
            self._fails.pop(target, None)
            self._opened_at.pop(target, None)
            self._persist_success(target)
            return
        count = self._fails.get(target, 0) + 1
        self._fails[target] = count
        if count >= self.threshold:
            self._opened_at[target] = self.clock()  # (re)open, restart the cooldown
        self._persist_failure(target)

    def _persist_failure(self, target: str) -> None:
        """Rewrite one target's row after a failure; best-effort, never raises."""
        if self._conn is None:
            return
        try:
            self._conn.execute(
                "INSERT INTO breaker(target, fails, opened_at) VALUES (?, ?, ?) "
                "ON CONFLICT(target) DO UPDATE SET fails=excluded.fails, opened_at=excluded.opened_at",
                (target, self._fails[target], self._opened_at.get(target)),
            )
        except (sqlite3.Error, OSError):
            self._conn = None

    def _persist_success(self, target: str) -> None:
        """Clear one target's row after a success; best-effort, never raises."""
        if self._conn is None:
            return
        try:
            self._conn.execute("DELETE FROM breaker WHERE target = ?", (target,))
        except (sqlite3.Error, OSError):
            self._conn = None


def delivery_plan(
    primary: str,
    fallbacks: Iterable[str],
    breaker: CircuitBreaker | None = None,
    allow: Callable[[str], bool] | None = None,
) -> list[str]:
    """Ordered, de-duplicated targets to try: the primary then its same-tier fallbacks.

    Targets whose breaker is open (still cooling down), or that ``allow`` rejects (a
    deterministic pre-call check, e.g. context-window fit), are dropped. Order and identity
    come from config, never from the score (WF-ADR-0031: delivery, not decision). May be
    empty if every candidate is filtered — the caller then fails fast.
    """
    plan: list[str] = []
    for target in (primary, *fallbacks):
        if target in plan:
            continue
        if breaker is not None and not breaker.allow(target):
            continue
        if allow is not None and not allow(target):
            continue
        plan.append(target)
    return plan


# Cross-tier failover policies (WF-ADR-0031). Default is conservative: stay on the chosen
# tier (only its configured alternate endpoints), changing neither cost nor answer quality.
FAILOVER_POLICIES = ("same-tier", "degrade", "escalate")


def failover_candidates(chosen: str, ladder: Iterable[str], policy: str) -> list[str]:
    """Cross-tier targets to try after same-tier endpoints are exhausted, per ``policy``.

    ``ladder`` is the tier model names cheapest→dearest. ``degrade`` walks to cheaper tiers
    (nearest-cheaper first; never raises cost); ``escalate`` walks to dearer tiers
    (nearest-dearer first; opt-in, raises cost); ``same-tier`` (and an off-ladder ``chosen``)
    yields nothing. Identity/order from the ladder, not the score.
    """
    seq = list(ladder)
    if policy not in ("degrade", "escalate") or chosen not in seq:
        return []
    idx = seq.index(chosen)
    if policy == "degrade":
        return seq[:idx][::-1]  # cheaper tiers, nearest first
    return seq[idx + 1:]  # escalate: dearer tiers, nearest first


def precheck_ok(estimated_tokens: int, context_window: int | None) -> bool:
    """A deterministic pre-call check: does the estimated prompt fit the target's window?

    ``None`` means no configured limit (always OK). Used to skip a target that would
    certainly fail on length before spending the call (WF-ADR-0031).
    """
    return context_window is None or estimated_tokens <= context_window

File: rust/crates/wayfinder-providers/src/reliability.rs
//! Deterministic provider-delivery policy.
//!
//! These functions change how an already-selected model is delivered.  They
//! must not change the routing decision itself.  Time and jitter are passed in
//! explicitly so compatibility tests never need a network or wall clock.

use std::collections::{HashMap, HashSet};

use wayfinder_core::python_round;

/// HTTP statuses that may succeed when attempted again.
pub const RETRYABLE_STATUS: [u16; 5] = [429, 500, 502, 503, 504];

/// HTTP statuses indicating an unusable upstream credential or proxy auth.
pub const AUTH_FAILURE_STATUS: [u16; 3] = [401, 403, 407];

/// Whether a provider attempt should be retried.
///
/// `None` represents a transport failure. Ordinary client errors fail fast.
#[must_use]
pub fn is_retryable(status: Option<u16>) -> bool {
    status.is_none_or(|value| RETRYABLE_STATUS.contains(&value))
}

/// Whether the provider target should count this response as an auth failure.
#[must_use]
pub fn is_auth_failure(status: Option<u16>) -> bool {
    status.is_some_and(|value| AUTH_FAILURE_STATUS.contains(&value))
}

/// Produce one full-jitter delay per retry.
///
/// The initial attempt has no delay. `rng` is called exactly once for every
/// retry and is expected to return a value in the inclusive range `0.0..=1.0`.
/// Values outside that range are clamped so a faulty jitter source cannot
/// create a negative or over-cap sleep.
#[must_use]
pub fn retry_delays_with(
    retries: usize,
    base_seconds: f64,
    cap_seconds: f64,
    mut rng: impl FnMut() -> f64,
) -> Vec<f64> {
    let base = if base_seconds.is_finite() {
        base_seconds.max(0.0)
    } else {
        0.0
    };
    let cap = if cap_seconds.is_finite() {
        cap_seconds.max(0.0)
    } else {
        0.0
    };

    (0..retries)
        .map(|index| {
            let exponent = i32::try_from(index).unwrap_or(i32::MAX);
            let slot = (base * 2_f64.powi(exponent)).min(cap);
            let sample = rng();
            let jitter = if sample.is_finite() {
                sample.clamp(0.0, 1.0)
            } else {
                0.0
            };
            python_round(slot * jitter, 6)
        })
        .collect()
}

/// A per-target circuit breaker whose clock is supplied by the caller.
///
/// Passing time explicitly avoids wall-clock behavior in tests and lets the
/// async gateway use Tokio's monotonic clock at its boundary. Timestamps are
/// opaque monotonic seconds; they must come from the same clock domain.
#[derive(Clone, Debug)]
pub struct CircuitBreaker {
    threshold: usize,
    cooldown_seconds: f64,
    failures: HashMap<String, usize>,
    opened_at: HashMap<String, f64>,
}

impl CircuitBreaker {
    /// Construct a breaker. A zero threshold behaves as one failure, and an
    /// invalid or negative cooldown behaves as no cooldown.
    #[must_use]
    pub fn new(threshold: usize, cooldown_seconds: f64) -> Self {
        Self {
            threshold: threshold.max(1),
            cooldown_seconds: if cooldown_seconds.is_finite() {
                cooldown_seconds.max(0.0)
            } else {
                0.0
            },
            failures: HashMap::new(),
            opened_at: HashMap::new(),
        }
    }

    /// Whether `target` may be tried now (closed or ready for a half-open probe).
    #[must_use]
    pub fn allow_at(&self, target: &str, now_seconds: f64) -> bool {
        self.opened_at
            .get(target)
            .is_none_or(|opened| now_seconds - *opened >= self.cooldown_seconds)
    }

    /// Whether the breaker is still cooling down for `target`.
    #[must_use]
    pub fn is_open_at(&self, target: &str, now_seconds: f64) -> bool {
        !self.allow_at(target, now_seconds)
    }

    /// Record the result of one attempt.
    ///
    /// Success closes the breaker. Every failure at or above the threshold
    /// restarts the cooldown, matching the Python oracle's failed-probe rule.
    pub fn record_at(&mut self, target: &str, succeeded: bool, now_seconds: f64) {
        if succeeded {
            self.failures.remove(target);
            self.opened_at.remove(target);
            return;
        }

        let count = self.failures.entry(target.to_owned()).or_default();
        *count = count.saturating_add(1);
        if *count >= self.threshold {
            self.opened_at.insert(target.to_owned(), now_seconds);
        }
    }
}

impl Default for CircuitBreaker {
    fn default() -> Self {
        Self::new(5, 30.0)
    }
}

/// Build an ordered, de-duplicated same-tier delivery plan.
///
/// Open targets and targets rejected by `allow` are removed. The result may be
/// empty, in which case the caller should fail before making a request.
pub fn delivery_plan<'a>(
    primary: &'a str,
    fallbacks: impl IntoIterator<Item = &'a str>,
    breaker: Option<(&CircuitBreaker, f64)>,
    mut allow: impl FnMut(&str) -> bool,
) -> Vec<String> {
    let mut seen = HashSet::new();
    std::iter::once(primary)
        .chain(fallbacks)
        .filter(|target| seen.insert((*target).to_owned()))
        .filter(|target| breaker.is_none_or(|(state, now)| state.allow_at(target, now)))
        .filter(|target| allow(target))
        .map(ToOwned::to_owned)
        .collect()
}

/// Cross-tier failover behavior after same-tier endpoints are exhausted.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub enum FailoverPolicy {
    /// Never move to a different tier.
    #[default]
    SameTier,
    /// Try cheaper tiers, nearest first.
    Degrade,
    /// Try more expensive tiers, nearest first.
    Escalate,
}

/// Return cross-tier candidates in compatibility order.
#[must_use]
pub fn failover_candidates(chosen: &str, ladder: &[String], policy: FailoverPolicy) -> Vec<String> {
    let Some(index) = ladder.iter().position(|model| model == chosen) else {
        return Vec::new();
    };

    match policy {
        FailoverPolicy::SameTier => Vec::new(),
        FailoverPolicy::Degrade => ladder[..index].iter().rev().cloned().collect(),
        FailoverPolicy::Escalate => ladder[index.saturating_add(1)..].to_vec(),
    }
}

/// Whether a prompt estimate fits the target's configured context window.
#[must_use]
pub fn precheck_ok(estimated_tokens: u64, context_window: Option<u64>) -> bool {
    context_window.is_none_or(|window| estimated_tokens <= window)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn retries_only_transport_and_transient_statuses() {
        assert!(is_retryable(None));
        for status in RETRYABLE_STATUS {
            assert!(is_retryable(Some(status)));
        }
        for status in [200, 400, 401, 403, 404, 422] {
            assert!(!is_retryable(Some(status)));
        }
    }

    #[test]
    fn auth_failure_set_matches_python() {
        for status in AUTH_FAILURE_STATUS {
            assert!(is_auth_failure(Some(status)));
        }
        assert!(!is_auth_failure(None));
        assert!(!is_auth_failure(Some(429)));
    }

    #[test]
    fn jitter_is_exponential_capped_and_rounded() {
        assert_eq!(retry_delays_with(4, 0.2, 1.0, || 1.0), [0.2, 0.4, 0.8, 1.0]);
        assert_eq!(retry_delays_with(3, 0.2, 5.0, || 0.0), [0.0, 0.0, 0.0]);
        assert!(retry_delays_with(0, 0.2, 5.0, || 1.0).is_empty());
    }

    #[test]
    fn jitter_source_is_contained() {
        assert_eq!(retry_delays_with(1, 1.0, 5.0, || 2.0), [1.0]);
        assert_eq!(retry_delays_with(1, 1.0, 5.0, || -1.0), [0.0]);
        assert_eq!(retry_delays_with(1, 1.0, 5.0, || f64::NAN), [0.0]);
    }

    #[test]
    fn breaker_opens_then_allows_probe_at_cooldown() {
        let mut breaker = CircuitBreaker::new(3, 30.0);
        breaker.record_at("cloud", false, 0.0);
        breaker.record_at("cloud", false, 0.0);
        assert!(breaker.allow_at("cloud", 0.0));
        breaker.record_at("cloud", false, 0.0);
        assert!(breaker.is_open_at("cloud", 29.0));
        assert!(breaker.allow_at("cloud", 30.0));
    }

    #[test]
    fn failed_probe_reopens_and_success_closes() {
        let mut breaker = CircuitBreaker::new(2, 10.0);
        breaker.record_at("x", false, 0.0);
        breaker.record_at("x", false, 0.0);
        assert!(breaker.allow_at("x", 10.0));
        breaker.record_at("x", false, 10.0);
        assert!(breaker.is_open_at("x", 19.0));
        assert!(breaker.allow_at("x", 20.0));
        breaker.record_at("x", true, 20.0);
        assert!(breaker.allow_at("x", 20.0));
    }

    #[test]
    fn breaker_is_per_target() {
        let mut breaker = CircuitBreaker::new(1, 999.0);
        breaker.record_at("a", false, 0.0);
        assert!(breaker.is_open_at("a", 0.0));
        assert!(breaker.allow_at("b", 0.0));
    }

    #[test]
    fn plan_orders_deduplicates_and_filters() {
        let mut breaker = CircuitBreaker::new(1, 999.0);
        breaker.record_at("cloud", false, 0.0);
        let plan = delivery_plan(
            "cloud",
            ["cloud", "cloud-2", "local"],
            Some((&breaker, 0.0)),
            |name| name != "local",
        );
        assert_eq!(plan, ["cloud-2"]);
    }

    #[test]
    fn plan_can_be_empty() {
        let mut breaker = CircuitBreaker::new(1, 999.0);
        breaker.record_at("cloud", false, 0.0);
        breaker.record_at("cloud-2", false, 0.0);
        let plan = delivery_plan("cloud", ["cloud-2"], Some((&breaker, 0.0)), |_| true);
        assert!(plan.is_empty());
    }

    #[test]
    fn failover_walks_nearest_tier_first() {
        let ladder = ["local", "mid", "cloud"].map(str::to_owned);
        assert_eq!(
            failover_candidates("mid", &ladder, FailoverPolicy::Degrade),
            ["local"]
        );
        assert_eq!(
            failover_candidates("local", &ladder, FailoverPolicy::Escalate),
            ["mid", "cloud"]
        );
        assert_eq!(
            failover_candidates("cloud", &ladder, FailoverPolicy::Degrade),
            ["mid", "local"]
        );
        assert!(failover_candidates("ghost", &ladder, FailoverPolicy::Degrade).is_empty());
    }

    #[test]
    fn precheck_honors_inclusive_context_limit() {
        assert!(precheck_ok(500, None));
        assert!(precheck_ok(500, Some(500)));
        assert!(!precheck_ok(501, Some(500)));
    }
}
